experimental/integrations/prototype-spark/marquez-spark-agent/src/main/java/marquez/spark/agent/lifecycle/SparkSQLExecutionContext.java

```diff
@@ -1,184 +1,206 @@
-    Collection<SparkPlanInfo> children = JavaConversions.asJavaCollection(i.children());
-    for (SparkPlanInfo child : children) {
-      print(prefix + "  ", child);
+  private List<Dataset> buildOutputs(LogicalPlan logical) {
+    List<Dataset> inputDatasets = new ArrayList<>();
+    if (logical instanceof InsertIntoHadoopFsRelationCommand) {
+      InsertIntoHadoopFsRelationCommand insert = (InsertIntoHadoopFsRelationCommand) logical;
+      inputDatasets.add(
+          Dataset.builder()
+              .namespace(marquezContext.getJobNamespace())
+              .name(insert.outputPath().toUri().toString())
+              .build());
     }
+    return inputDatasets;
   }
 
-  private void print(String prefix, SparkPlan executedPlan) {
-    extraInfo.append(prefix + executedPlan.treeString(true, true)).append("\n");
+  private LineageEvent.Run buildRun(RunFacet facets) {
+    return LineageEvent.Run.builder().runId(marquezContext.getParentRunId()).facets(facets).build();
   }
 
-  private void print(String prefix, LogicalPlan plan) {
-    String extra = "";
-    if (plan instanceof LogicalRelation) {
-      LogicalRelation lr = (LogicalRelation) plan;
-      BaseRelation lrr = lr.relation();
-      extra = "||| " + lrr + " " + lr.output() + " " + lr.catalogTable() + " " + lr.isStreaming();
-      if (lrr instanceof HadoopFsRelation) {
-        FileIndex location = ((HadoopFsRelation)lrr).location();
-        extra += "|||" + Arrays.toString(location.inputFiles());
-      }
+  protected RunFacet buildRunFacets(
+      JsonNode logicalPlanFacet, JsonNode physicalPlanFacet, Object jobError) {
+    Map<String, Object> additionalFacets = new HashMap<>();
+    if (logicalPlanFacet != null) {
+      additionalFacets.put(LogicalPlan.class.getName(), logicalPlanFacet);
+    }
+    if (physicalPlanFacet != null) {
+      additionalFacets.put(SparkPlan.class.getName(), physicalPlanFacet);
     }
-    extraInfo.append(prefix + plan.getClass().getSimpleName() + ":" + extra + " plan: " + plan).append("\n");
-    Collection<LogicalPlan> children = JavaConversions.asJavaCollection(plan.children());
-    for (LogicalPlan child : children) {
-      print(prefix + "  ", child);
+    if (jobError != null) {
+      additionalFacets.put("spark.exception", jobError);
     }
+    return RunFacet.builder().additional(additionalFacets).build();
   }
 
-  public void end(SparkListenerSQLExecutionEnd endEvent) {
-    logger.info("Spark sql execution ended " + endEvent);
-    if (success == null) {
-      extraInfo.append("unknown end status\n");
-    } else if (success) {
-      marquezContext.success(runId, jobId, endEvent.time(), outputs);
-    } else  {
-      marquezContext.failure(runId, jobId, endEvent.time(), outputs, error );
+  private JsonNode buildLogicalPlanFacet(LogicalPlan plan) {
+    try {
+      return OpenLineageClient.getObjectMapper().readTree(plan.toJSON());
```

Comment:
Do you have an example of what this looks like and whether it contains the inputs an outputs details for example?

Reply:
On second look, this json output is more limited than I anticipated. I wrote up some code that traverses the logical plan but I abandoned it because I thought the json plan included objects with more granularity. I'll pull that code back out.

Reply:
example of the toJSON output: https://gist.github.com/henneberger/081e4a615214b2bbd15f939738ca4ded